doc/LAYERS.org

```diff
@@ -0,0 +1,540 @@
+your file is actually evaluated, probably not what was intended:
+
+#+begin_src emacs-lisp
+  ;;;###autoload
+  (require 'helm)
+#+end_src
+
+It is the responsibility of the package authors to ensure that their package can
+be appropriately auto-loaded, and most packages do this quite well.
+
+Spacemacs makes thorough use of auto-loading. Almost everything in Spacemacs is
+loaded when needed instead of right away.
+
+** Use-package
+For /end users/ who are trying to put together an efficient Emacs configuration,
+there is a very useful /package/ called use-package that provides a macro called
+=use-package= which does a very good job of streamlining the whole process of
+loading packages and extensions.
+
+The aspiring layer author is recommended to have a look at the =use-package=
+[[https://github.com/jwiegley/use-package][documentation]]. Some examples follow.
+
+#+begin_src emacs-lisp
+  (use-package helm)
+#+end_src
+
+This simply loads Helm. It is essentially equivalent to =(require 'helm)=.
+
+#+begin_src emacs-lisp
+  (use-package helm
+    :defer t)
+#+end_src
+
+This defers the loading of Helm using the auto-load facility and the auto-load
+commands provided by the Helm source code.
+
+#+begin_src emacs-lisp
+  (use-package helm
+    :defer t
+    :init
+    ;; Code to execute before Helm is loaded
+    :config
+    ;; Code to execute after Helm is loaded
+    )
+#+end_src
+
+This form includes code to execute before and after Helm is loaded. The =:init=
+section can be executed immediately, but since Helm is deferred, the =:config=
+section is not executed until after loading, if ever.
+
+#+begin_src emacs-lisp
+  (use-package helm
+    :commands (helm-find-files helm-M-x))
+#+end_src
+
+This creates auto-load references for additional commands, if you find that the
+package author has been slacking.
+
+#+begin_src emacs-lisp
+  (use-package ruby-mode
+    :mode "\\.rb\\'")
+#+end_src
+
+For packages that provide major modes, you can associate file extensions to that
+mode by using the =:mode= keyword. This adds an entry to =auto-mode-alist= and
+an auto-load for =ruby-mode=. Typically this is not required, as =ruby-mode=
+should already be auto-loadable, and the package should associate Ruby files
+with itself already.
+
+Use-package supports heaps of useful keywords. Look at the [[https://github.com/jwiegley/use-package][documentation]] for
+more.
+
+* Anatomy of a layer
+A layer is simply a folder somewhere in Spacemacs' layer search path that
+contains /at least one/ of these files, ordered vaguely by importance.
+
+- =packages.el= :: for specifying which packages the layer uses
+- =extensions.el= :: for specifying extensions
+- =config.el= :: for miscellaneous configuration
+- =funcs.el= :: for defining general utility functions
+- =keybindings.el= :: for defining general keybindings
+
+Additionally, for each extension, there should be a folder
+=<layer>/extensions/<extension>/= containing the source code for that extension.
```

Comment:
The folder will be renamed to "local" instead of "extensions". The new name will become mandatory in 0.105.0 (in about 2 months).